net/server_encryption_key.go

```diff
@@ -0,0 +1,351 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package net
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"fmt"
+
+	"encoding/base64"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+	libpeer "github.com/libp2p/go-libp2p/core/peer"
+	rpc "github.com/sourcenetwork/go-libp2p-pubsub-rpc"
+	"github.com/sourcenetwork/immutable"
+	grpcpeer "google.golang.org/grpc/peer"
+	"google.golang.org/protobuf/proto"
+
+	libp2pCrypto "github.com/libp2p/go-libp2p/core/crypto"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/crypto"
+	"github.com/sourcenetwork/defradb/errors"
+	"github.com/sourcenetwork/defradb/internal/core"
+	"github.com/sourcenetwork/defradb/internal/encryption"
+	pb "github.com/sourcenetwork/defradb/net/pb"
+)
+
+const encryptionTopic = "encryption"
+
+// getEncryptionKeys retrieves the encryption keys for the given targets.
+// It returns the encryption keys and the targets for which the keys were found.
+func (s *server) getEncryptionKeys(
+	ctx context.Context,
+	req *pb.FetchEncryptionKeyRequest,
+) ([]byte, []*pb.EncryptionKeyTarget, error) {
+	encryptionKeys := make([]byte, 0, len(req.Targets))
+	targets := make([]*pb.EncryptionKeyTarget, 0, len(req.Targets))
+	for _, target := range req.Targets {
+		docID, err := client.NewDocIDFromString(string(target.DocID))
+		if err != nil {
+			return nil, nil, err
+		}
+
+		optFieldName := immutable.None[string]()
+		if target.FieldName != "" {
+			optFieldName = immutable.Some(target.FieldName)
+		}
+		encKey, err := encryption.GetKey(
+			encryption.ContextWithStore(ctx, s.peer.encstore),
+			core.NewEncStoreDocKey(docID.String(), optFieldName, string(target.KeyID)),
+		)
+		if err != nil {
+			return nil, nil, err
+		}
+		// TODO: we should test it somehow. For this this one peer should have some keys and
+		// another one should have the others. https://github.com/sourcenetwork/defradb/issues/2895
+		if len(encKey) == 0 {
+			continue
```

Comment:
praise: Thanks for leaving this comment here and the issue is quite descriptive as well.